src/Icicle/Core/Exp/Simp.hs

```diff
@@ -29,40 +29,39 @@ import qualified Data.Set                       as Set
 --   * ...something exciting???
 --
 simp :: Ord n => a -> (C.Exp a n -> Bool) -> C.Exp a n -> Fresh n (C.Exp a n)
-simp a_fresh isValue = anormal a_fresh . deadX . fixp (50 :: Int) (simpX a_fresh isValue)
+simp a_fresh isValue = anormal a_fresh . deadX . fixp (simpX a_fresh isValue)
  where
-  fixp 0 f = f
-  fixp n f = f . fixp (n-1) f
+  fixp f x
+   | Just x' <- f x = fixp f x'
+   | otherwise      = x
 
 
-simpX :: Ord n => a -> (C.Exp a n -> Bool) -> C.Exp a n -> C.Exp a n
+simpX :: Ord n => a -> (C.Exp a n -> Bool) -> C.Exp a n -> Maybe (C.Exp a n)
 simpX a_fresh isValue = go . beta
   where
     beta  = B.beta isValue
     go xx = case xx of
       -- * constant folding for some primitives
       XApp a p q
-        | p' <- go p
-        , q' <- go q
+        | Just p' <- go p
+        , Just q' <- go q
```

Comment:
won't this only happen if both sub-expressions succeed? but they won't succeed if they're base cases?